src/primitives.jl

```diff
@@ -462,3 +491,161 @@ function draw_atomic(scene::Scene, screen::CairoScreen, primitive::AbstractPlott
+        Cairo.mesh_pattern_begin_patch(pattern)
+
+        Cairo.mesh_pattern_move_to(pattern, t1[1], t1[2])
+        Cairo.mesh_pattern_line_to(pattern, t2[1], t2[2])
+        Cairo.mesh_pattern_line_to(pattern, t3[1], t3[2])
+
+        mesh_pattern_set_corner_color(pattern, 0, c1)
+        mesh_pattern_set_corner_color(pattern, 1, c2)
+        mesh_pattern_set_corner_color(pattern, 2, c3)
+
+        Cairo.mesh_pattern_end_patch(pattern)
+    end
+    Cairo.set_source(ctx, pattern)
+    Cairo.close_path(ctx)
+    Cairo.paint(ctx)
+    return nothing
+end
+
+
+################################################################################
+#                                   Surface                                    #
+################################################################################
+
+function draw_atomic(scene::Scene, screen::CairoScreen, primitive::AbstractPlotting.Surface)
+    # Pretend the surface plot is a mesh plot and plot that instead
+    mesh = surface2mesh(primitive[1][], primitive[2][], primitive[3][])
+    primitive[:color][] = primitive[3][][:]
+    draw_mesh3D(scene, screen, primitive, mesh)
+    return nothing
+end
+
+function surface2mesh(xs::Vector, ys::Vector, zs::Matrix)
+    ps = [Point3f0(xs[i], ys[j], zs[i, j]) for j in eachindex(ys) for i in eachindex(xs)]
+    idxs = LinearIndices(size(zs))
+    faces = [
+        QuadFace(idxs[i, j], idxs[i+1, j], idxs[i+1, j+1], idxs[i, j+1]) 
+        for j in 1:size(zs, 2)-1 for i in 1:size(zs, 1)-1
+    ]
+    normal_mesh(ps, faces)
+end
+function surface2mesh(xs::Matrix, ys::Matrix, zs::Matrix)
+    ps = [Point3f0(xs[i, j], ys[i, j], zs[i, j]) for j in 1:size(zs, 2) for i in 1:size(zs, 1)]
+    idxs = LinearIndices(size(zs))
+    faces = [
+        QuadFace(idxs[i, j], idxs[i+1, j], idxs[i+1, j+1], idxs[i, j+1]) 
+        for j in 1:size(zs, 2)-1 for i in 1:size(zs, 1)-1
+    ]
+    normal_mesh(ps, faces)
+end
+    
+
+################################################################################
+#                                 MeshScatter                                  #
+################################################################################
+
+function draw_atomic(scene::Scene, screen::CairoScreen, primitive::AbstractPlotting.MeshScatter)
+    m = normal_mesh(primitive[:marker][])
+    pos = primitive[1][]
+    # For correct z-ordering we need to be in view/camera or screen space
+    model = primitive[:model][]
+    view = scene.camera.view[]
+    sort!(pos, by = p -> begin
+        p4d = to_ndim(Vec4f0, to_ndim(Vec3f0, p, 0f0), 1f0)
+        cam_pos = view * model * p4d
+        cam_pos[3] / cam_pos[4]
+    end, rev=false)
+
+    colors = primitive[:color][]
+
+    for i in eachindex(pos)
+        p = pos[i]
+        if colors isa Vector
+            primitive[:color][] = colors[i]
+        end
+        draw_mesh3D(scene, screen, primitive, m, p, primitive[:markersize][])
+    end
+
+    primitive[:color][] = colors
+    return nothing
+end
```

Comment:
I thought about creating new `Mesh` plots here (and in the method for surface) since that's the primitive I'm drawing in the end. What would be the way to do that? 

I tried copying stuff directly, i.e. `Mesh(parent, transformation, attributes, input_args, converted)`, but that needs type information (`Mesh{...}`) which I haven't figure out yet.

Reply:
The loop seems pretty fair?  I'm not sure if it would be any easier to create a mesh, which would also need a very similar loop?

Reply:
I don't feel good about modifying attributes of the input. I just have `:color` here, but I assume there are few more that I'm missing. I could also pass all of those to `draw_mesh3D ` directly, which would probably be the easiest solution.

Creating a new Mesh object seems like the most readable/conceptually straight forward to me. `Surface` and `MeshScatter` objects would be transformed to `Mesh` objects and then just drawn without any extra fuzz. But maybe that's just me getting pedantic...